dictionary.go

```diff
@@ -709,36 +711,46 @@ func (d *byteArrayDictionary) Insert(indexes []int32, values []Value) {
 	d.insert(indexes, makeArrayValue(values, unsafe.Offsetof(model.ptr)))
 }
 
-func (d *byteArrayDictionary) insert(indexes []int32, rows sparse.Array) {
-	_ = indexes[:rows.Len()]
+func (d *byteArrayDictionary) init() {
+	d.table = make(map[string]int32, d.numValues)
 
-	if d.hashmap == nil {
-		d.hashmap = make(map[string]int32, cap(d.offsets))
-		for index, offset := range d.offsets {
-			d.hashmap[string(d.valueAt(offset))] = int32(index)
-		}
+	for i := 0; i < len(d.values); {
+		n := plain.ByteArrayLength(d.values[i:])
+		i += plain.ByteArrayLengthSize
+		s := d.values[i : i+n]
+		i += n
+		d.table[string(s)] = int32(len(d.table))
 	}
+}
 
-	for i := 0; i < rows.Len(); i++ {
-		value := *(*string)(rows.Index(i))
+func (d *byteArrayDictionary) insert(indexes []int32, rows sparse.Array) {
+	if d.table == nil {
+		d.init()
+	}
 
-		index, exists := d.hashmap[value]
+	values := rows.StringArray()
+
+	for i := range indexes {
+		value := values.Index(i)
+
+		index, exists := d.table[value]
 		if !exists {
-			index = int32(len(d.offsets))
-			value = d.append(value)
-			d.hashmap[value] = index
+			value = cloneString(value)
+			index = d.numValues
+			d.numValues++
+			d.table[value] = index
+			d.offsets = append(d.offsets, uint32(len(d.values)))
+			d.values = plain.AppendByteArrayString(d.values, value)
 		}
 
 		indexes[i] = index
 	}
 }
 
-func (d *byteArrayDictionary) append(value string) string {
-	offset := len(d.values)
-	d.values = plain.AppendByteArrayString(d.values, value)
-	d.offsets = append(d.offsets, uint32(offset))
-	d.numValues++
-	return string(d.values[offset+plain.ByteArrayLengthSize : len(d.values)])
+func cloneString(s string) string {
+	b := new(strings.Builder)
+	b.WriteString(s)
+	return b.String()
```

Comment:
This can't be the best way to do this, can it...?

Reply:
sigh... 

```
// String returns the accumulated string.
func (b *Builder) String() string {
	return *(*string)(unsafe.Pointer(&b.buf))
}

Reply:
Go 1.18 has [`strings.Clone](https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/strings/clone.go;l=21-28) but we have to maintain compatibility with Go 1.17 for now.